feat: Add cache layer for the datastore adapter

ArneGudermann opened the pull request.

ArneGudermann committed:
Add cache layer for the datastore

ArneGudermann committed:
PEP 8

ArneGudermann committed:
Rework delete

phorward commented on src/viur/core/db/transport.py:
```diff
@@ -83,10 +99,23 @@ def delete(keys: t.Union[Entity, t.List[Entity], Key, t.List[Key]]):
-        return __client__.delete(keys)
-
-    return __client__.delete_multi(keys)
+    keys_data = []
+    single_request = isinstance(keys, (Key, Entity))
+    if isinstance(keys, Entity):
+        keys_data = keys.key
+    elif isinstance(keys, Key):
+        keys_data = keys
+    elif not single_request:
+        for key in keys:
+            if isinstance(key, Key):
+                keys_data.append(key)
+            elif isinstance(key, Entity):
+                keys_data.append(key.key)
+    cache.delete(keys_data)
+    if single_request:
+        return __client__.delete(keys_data)
+
+    return __client__.delete_multi(keys_data)
```
Same here. Assume keys is always an iterable.

```suggestion
    keys = (k.key if isinstance(k, Entity) else k for k in utils.ensure_iterable(keys))
    if not keys:
        return
    
    cache.delete(keys)
    if len(keys) == 1:
        return __client__.delete(keys)

    return __client__.delete_multi(keys)
```

ArneGudermann replied:
@phorward I use get and not get_multi because it's faster.

phorward reviewed (changes requested):
Thanks for facing this.

I'm a little bit astonished how bloated this code is, especially regarding this multiple-key feature. Is this anywhere used in ViUR, except in legacy code? 

Anyway. I think this code might be improved in the cache-module, too.

My recommendation: When code that either handles one or a list of keys is written, write the code as it always handles a list of keys, so that the `len(keys) == 1`-case becomes the special case. This eliminates lots of bloating regarding isinstance()-checks and iterable-checking.

Can you please test my suggestions, and please don't forget to import `utils` to use `ensure_iterable`.

ArneGudermann committed:
Update src/viur/core/db/transport.py

Co-authored-by: Jan Max Meyer <jmm@phorward.de>

ArneGudermann committed:
Rework db.get

phorward committed:
Merge branch 'develop' into feat/db-cache

phorward committed:
fix: Move `utils.parse.sortindex` into `db.SortIndex.from_str`

The previous solution was because SortIndex was in viur-datastore previously.

phorward committed:
fix: Use `utils.ensure_iterable()`

phorward commented:
> @phorward we can not use `utils` due a circular import

I've fixed that and also a a bug and update the code to use `utils.ensure_iterable()` can you please have a look and - if you have test-cases - test it?

ArneGudermann committed:
Merge branch 'develop' into feat/db-cache

ArneGudermann committed:
Remove old test

ArneGudermann committed:
feat: bypass datastore cache within transactions

cache.get() and cache.put() now short-circuit while a datastore
transaction is active, so reads go straight to the datastore and no
uncommitted (potentially rolled-back) writes get cached.

Also wire the cache layer into transport.get(): serve hits from the
cache, fetch missing keys via get_multi and write them back. Fixes the
single-key path (UnboundLocalError), set inputs and the list/entity
return contract along the way.

ArneGudermann committed:
refactor: drop redundant local db import in normalizeKey

db is already imported at module level, so the function-local import is
unnecessary.
